Source/JavaScriptCore/runtime/ImportMap.cpp

```diff
@@ -254,18 +284,189 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+        // module set, we have a scoped resolved module map. The map's keys are
+        // scope prefixes, and its values are a set of specifier prefixes that
+        // already exist in that scope. We grab the set of specifier prefixes using
+        // the current scope and then iterate over the scope's imports, removing any
+        // specifiers whose prefix is in the set.
+        auto iter = m_scopedResolvedModuleMap.find(AtomString(scope.key.string()));
+        if (iter != m_scopedResolvedModuleMap.end()) {
+            auto& currentResolvedSet = iter->value;
+            Vector<AtomString> specifiersToRemove;
+            for (auto& specifier : scopeImports.keys()) {
+                if (currentResolvedSet.find(specifier) != currentResolvedSet.end())
+                    specifiersToRemove.append(specifier);
+            }
+            for (auto& specifier : specifiersToRemove) {
+                // 3.1.2.1.1. The user agent may report the removed rule as a warning to
+                // the developer console.
+                reporter.reportWarning(makeString("An import map scope rule for specifier '"_s, specifier, "' was removed, as it conflicted with already resolved module specifiers."_s));
+                // 3.1.2.1.2. Remove scopeImports[specifier].
+                scopeImports.remove(specifier);
+            }
+        }
+
+        // 3.2 If scopePrefix exists in oldImportMap's scopes, then set
+        // oldImportMap's scopes[scopePrefix] to the result of merging module
+        // specifier maps, given scopeImports and oldImportMap's
+        // scopes[scopePrefix].
+        const auto oldScopeSpecifierMapIt = m_scopesMap.find(scope.key);
+        if (oldScopeSpecifierMapIt != m_scopesMap.end()) {
+            ImportMap::SpecifierMap& oldScopeSpecifierMap = oldScopeSpecifierMapIt->value;
+            ImportMapInternal::mergeModuleSpecifierMaps(oldScopeSpecifierMap, scopeImports, reporter);
+        } else {
+            // 3.3 Otherwise, set oldImportMap's scopes[scopePrefix] to
+            // scopeImports.
+            m_scopesMap.set(scope.key, WTFMove(scopeImports));
+            m_scopesVector.append(scope.key);
+        }
+    }
+
+    // 4. For each url → integrity of newImportMap's integrity:
+    for (auto& url : newImportMapIntegrity.keys()) {
+        const auto& newIntegrityValue = newImportMapIntegrity.get(url);
+        // 4.1 If url exists in oldImportMap's integrity, then:
+        if (m_integrity.contains(url)) {
+            // 4.1.1. The user agent may report the removed rule as a warning to the
+            // developer console.
+            reporter.reportWarning(makeString("An import map integrity rule for url '"_s, url.string(), "' was removed, as it conflicted with already defined integrity rules."_s));
+            // 4.1.2 Continue.
+            continue;
+        }
+        // 4.2 Set oldImportMap's integrity[url] to integrity.
+        m_integrity.set(url, newIntegrityValue);
```

Comment:
Use `auto iter = m_integrity.add(url, newIntegrityValue)`. It will add that value only when `m_integrity` does not have that. And by checking `iter.isNewEntry`, you can detect whether it was containing the entry before. (So, you can avoid double-lookup here).